File: framework/ruby/src/soda/unittest/validate.rb
require_relative 'feature-factory'

module Soda end

module Soda::Unittest

class Validators
  def self.for_array(obj_type, ordered)
    _build_for_array(ordered, FeatureFactory.create(obj_type))
  end

  def self.for_array2d(obj_type, dim1_ordered, dim2_ordered)
    elem_feat = FeatureFactory.create(obj_type)
    d = dim2_ordered ? ListFeature.new(elem_feat) : UnorderedListFeature.new(elem_feat)
    _build_for_array(dim1_ordered, d)
  end

  def self._build_for_array(ordered, elem_feat)
    ordered ? ->(a,b) { ListFeature.new(elem_feat).is_equal?(a,b) }
      : ->(a,b) { UnorderedListFeature.new(elem_feat).is_equal?(a,b) }
  end
end

end  # module Soda::Unittest

File: framework/ts/src/soda/unittest/validate.ts
import {FeatureFactory} from "./featurefactory";
import {ObjectFeature} from "./feature";
import {ListFeatureFactory} from "./listfeature";

export class Validators {
    static forArray<T>(objType: string, ordered: boolean): (x: T[], y: T[]) => boolean {
        return this.for_arr<T>(ordered, FeatureFactory.create<T>(objType));
    }

    static forArray2d<T>(
        objType: string,
        dim1Ordered: boolean,
        dim2Ordered: boolean): (x: T[][], y: T[][]) => boolean
    {
        let elemFeat = FeatureFactory.create(objType);
        let d = dim2Ordered ? ListFeatureFactory.ordered<T>(elemFeat) : ListFeatureFactory.unordered<T>(elemFeat);
        return this.for_arr(dim1Ordered, d);
    }

    static for_arr<T>(ordered: boolean, elemFeat: ObjectFeature<T>): (x: T[], y: T[]) => boolean {
        if (ordered) {
            return (x: T[], y: T[]) => ListFeatureFactory.ordered<T>(elemFeat).isEqual(x, y);
        }
        return (x: T[], y: T[]) => ListFeatureFactory.unordered<T>(elemFeat).isEqual(x, y);
    }
}